File: libvitriol/pymander.py
#!/usr/bin/env python3
"""Pymander — the reference mind: static, curated, domain-specific knowledge.

Pymander is VITRIOL's *static* knowledge basis — *what we are / how we do a
domain well* — as opposed to Hermetis (episodic memory: *what happened*).
Content is hand-authored **atomic nodes** (small pieces of relevant knowledge
tied to the thing being written). This is the P1 store + ingest path only:
doctrine injection into the window and the `pymander` opencode tool arrive in
P2.

Store: each domain is a distinct Hermetis memory root
`~/.vitriol/pymander/<domain>/memory.db`, reached through the existing
`hermetis.db` machinery (project_id = `pymander/<domain>`), so node
versioning (git_rev supersede), embedding cache, and strength all come free.

Ingest format (markdown): `## Heading` starts an atomic node; the body up to
the next `##` is its summary. `#` titles and prose before the first `##` are
domain metadata and are skipped.

PROVENANCE: user-repo — VITRIOL's own architecture (Hermetis db/retrieval),
re-derived from this project's docs; no third-party source consulted.
"""
import argparse
import json
import os
import re
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from hermetis import db, retrieval

DOMAIN_PREFIX = "pymander/"
DOMAIN_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")


def _selection_path():
    """JSON file mapping project_id -> [active domain names]."""
    return os.path.join(db.MEMORY_DIR, "pymander", "selection.json")


def _load_selection() -> dict:
    """Read selection.json (missing/corrupt -> empty)."""
    try:
        with open(_selection_path(), "r", encoding="utf-8") as fh:
            data = json.load(fh)
            return data if isinstance(data, dict) else {}
    except (FileNotFoundError, json.JSONDecodeError, OSError):
        return {}


def _save_selection(data: dict):
    """Atomically write selection.json."""
    path = _selection_path()
    os.makedirs(os.path.dirname(path), exist_ok=True)
    tmp = path + ".tmp"
    with open(tmp, "w", encoding="utf-8") as fh:
        json.dump(data, fh, indent=2, sort_keys=True)
    os.replace(tmp, path)


def sanitize_domain(domain: str) -> str:
    """Validate + normalize a domain name.

    Domain becomes a path component under MEMORY_DIR/pymander/, so it must be a
    safe filesystem name. Raises ValueError on anything else.
    """
    if not DOMAIN_RE.match(domain):
        raise ValueError(
            f"invalid domain {domain!r}: use letters/digits/._- , no slashes"
        )
    return domain


def domain_project_id(domain: str) -> str:
    """Hermetis project_id backing a Pymander domain (pymander/<domain>)."""
    return DOMAIN_PREFIX + sanitize_domain(domain)


def list_domains() -> list[str]:
    """Return the installed Pymander domains (dirs holding a memory.db)."""
    root = os.path.join(db.MEMORY_DIR, "pymander")
    if not os.path.isdir(root):
        return []
    out = []
    for name in sorted(os.listdir(root)):
        full = os.path.join(root, name)
        if os.path.isdir(full) and os.path.exists(os.path.join(full, "memory.db")):
            out.append(name)
    return out


def _parse_markdown(md_text: str) -> list[tuple[str, str]]:
    """Split markdown into (label, summary) atomic nodes.

    A `## Heading` opens a node; its body (until the next `##` or end) is the
    summary. Content before the first `##` is treated as domain header and
    skipped. Empty summaries are dropped.
    """
    nodes: list[tuple[str, str]] = []
    label = None
    body: list[str] = []
    for line in md_text.splitlines():
        if line.startswith("## "):
            _flush_node(nodes, label, body)
            label = line[3:].strip()
            body = []
        elif label is not None:
            body.append(line)
    _flush_node(nodes, label, body)
    return nodes


def _flush_node(nodes: list[tuple[str, str]], label, body: list[str]):
    """Append a completed node when it has a label and non-empty summary."""
    if label is None:
        return
    summary = "\n".join(body).strip()
    if summary:
        nodes.append((label, summary))


def _embed_best_effort(project_id: str, summary: str):
    """Cache an embedding for a node summary when the embed server is up.

    Best-effort only: a missing/unreachable embed server must never fail an
    ingest (keyword scoring remains the fallback in retrieval).
    """
    try:
        from hermetis import embed
        if not embed.is_available():
            return
        vec = embed.encode(summary)
        if vec is None:
            return
        import numpy as np
        blob = np.array(vec, dtype="float32").tobytes()
        ch = db._content_hash(summary)
        conn = db._get_conn(project_id)
        db._store_cached_embedding(conn, ch, "pymander_node", blob)
        conn.commit()
    except Exception:
        pass


def ingest_markdown(domain: str, md_text: str, git_rev: str = "") -> dict:
    """Ingest a markdown corpus as atomic nodes for a domain.

    Returns {domain, stored, refreshed, embedded}. Same (label, git_rev) ->
    refresh in place; new git_rev -> supersede (never hard-discard), matching
    db.store_node's versioned semantics.
    """
    domain = sanitize_domain(domain)
    pid = domain_project_id(domain)
    nodes = _parse_markdown(md_text)
    conn = db._get_conn(pid)
    stored = 0
    refreshed = 0
    embedded = 0
    for label, summary in nodes:
        # 2026-08-07: detect refresh by checking the (label, git_rev) row that
        # store_node will update in place vs a fresh insert (new version).
        existing = conn.execute(
            "SELECT 1 FROM knowledge_nodes WHERE label=? AND git_rev=?",
            (label, git_rev)).fetchone()
        db.store_node(pid, label, summary,
                      meta={"git_rev": git_rev, "strength": 1.0})
        if existing:
            refreshed += 1
        else:
            stored += 1
        _embed_best_effort(pid, summary)
        if _has_embedding(pid, summary):
            embedded += 1
    return {"domain": domain, "nodes": len(nodes),
            "stored": stored, "refreshed": refreshed, "embedded": embedded}


def _has_embedding(project_id: str, summary: str) -> bool:
    """True if a cached embedding row exists for the summary."""
    try:
        ch = db._content_hash(summary)
        conn = db._get_conn(project_id)
        return conn.execute(
            "SELECT 1 FROM embeddings WHERE content_hash=?",
            (ch,)).fetchone() is not None
    except Exception:
        return False


def list_nodes(domain: str) -> list[dict]:
    """Return the current (superseded=0) nodes of a domain."""
    pid = domain_project_id(domain)
    return db.search_nodes(pid, "", limit=100000)


def search(domain: str, query: str, top_k: int = 5) -> list[dict]:
    """Retrieve the most relevant nodes of a domain for a query."""
    pid = domain_project_id(domain)
    candidates = retrieval.retrieve(pid, query, top_k=top_k, cascade_depth=0)
    return [c for c in candidates if c.get("_type") == "node"]


def set_selection(project_id: str, domains: list[str]) -> dict:
    """Set the active Pymander domains for a project (persisted)."""
    data = _load_selection()
    clean = [sanitize_domain(d) for d in domains]
    data[project_id] = clean
    _save_selection(data)
    return {"project_id": project_id, "domains": clean}


def get_selection(project_id: str) -> list[str]:
    """Return the active Pymander domains for a project (empty if unset)."""
    return list(_load_selection().get(project_id, []))


def _read_file_or_stdin(path: str) -> str:
    """Read a corpus file, or stdin when path is '-'.

    2026-08-07: '-' allows `cat corpus.md | vitriol pymander ingest dom -`,
    so authoring can pipe from $EDITOR or generated content.
    """
    if path == "-":
        return sys.stdin.read()
    with open(path, "r", encoding="utf-8") as fh:
        return fh.read()


def _cmd_ingest(args) -> int:
    domain = sanitize_domain(args.domain)
    md = _read_file_or_stdin(args.file)
    git_rev = args.rev if args.rev is not None else _repo_rev(args.file)
    res = ingest_markdown(domain, md, git_rev)
    print(json.dumps(res, indent=2, sort_keys=True))
    return 0


def _repo_rev(path: str) -> str:
    """Best-effort git HEAD of the corpus file's repo ('' when not in a repo)."""
    if path == "-" or not os.path.isfile(path):
        return ""
    try:
        import subprocess
        base = os.path.dirname(os.path.abspath(path))
        out = subprocess.run(
            ["git", "-C", base, "rev-parse", "HEAD"],
            capture_output=True, text=True, timeout=5)
        return out.stdout.strip() if out.returncode == 0 else ""
    except Exception:
        return ""


def _cmd_list(_args) -> int:
    domains = list_domains()
    print(json.dumps(domains, indent=2, sort_keys=True))
    return 0


def _cmd_nodes(args) -> int:
    nodes = list_nodes(args.domain)
    print(json.dumps(nodes, indent=2, sort_keys=True, default=str))
    return 0


def _cmd_search(args) -> int:
    res = search(args.domain, args.query, args.top_k)
    print(json.dumps(res, indent=2, sort_keys=True, default=str))
    return 0


def _cmd_select(args) -> int:
    res = set_selection(args.project_id, args.domains)
    print(json.dumps(res, indent=2, sort_keys=True))
    return 0


def _cmd_active(args) -> int:
    print(json.dumps(get_selection(args.project_id), sort_keys=True))
    return 0


def build_doctrine(project_id: str, query: str = "", budget_tokens: int = 3000,
                   top_k: int = 3) -> str:
    """Build a budgeted doctrine block for the project's selected domains.

    Aggregates the top nodes of each selected domain into a labeled, budgeted
    text block intended for window injection (the [Pymander doctrine] context).
    The model's static *how* lives here, selected per project. Falls back to the
    first installed domain when the project has no explicit selection.
    """
    domains = get_selection(project_id) or list_domains()[:1]
    sections = []
    used = 0
    for domain in domains:
        text = _domain_section(domain, query, top_k, budget_tokens - used)
        if not text:
            continue
        text_toks = retrieval.estimate_tokens(text) + 1
        if used + text_toks > budget_tokens and used > 0:
            break
        used += text_toks
        sections.append(text)
    return "\n\n".join(sections)


def _domain_section(domain: str, query: str, top_k: int, budget: int) -> str:
    """One domain's doctrine lines under a per-domain budget ('' when none)."""
    try:
        hits = search(domain, query, top_k=top_k)
    except ValueError:
        return ""
    if not hits:
        return ""
    parts = [f"## {domain}"]
    used = 0
    for hit in hits:
        body = f"- {hit.get('label', '')}: {hit.get('summary', '')}"
        toks = retrieval.estimate_tokens(body) + 1
        if used + toks > budget and used > 0:
            break
        used += toks
        parts.append(body)
    return "\n".join(parts)


def _candidates_path() -> str:
    """JSON file listing curated Ascensus/Hermetis answers worth promoting."""
    return os.path.join(db.MEMORY_DIR, "pymander", "candidates.json")


def add_candidate(domain: str, label: str, summary: str, source: str = "") -> dict:
    """Add a promotion candidate (curated: a user/machine decision to fold in).

    Candidates are NOT auto-merged into the corpus — the user reviews them and
    either re-ingests via `vitriol pymander ingest` or deletes. This is the
    learning loop's second half (Ascensus → Hermetis → candidate → curated
    promotion), implemented per AGENTS cleanroom (quality + license care).
    """
    domain = sanitize_domain(domain)
    path = _candidates_path()
    try:
        with open(path, "r", encoding="utf-8") as fh:
            data = json.load(fh)
        if not isinstance(data, dict):
            data = {}
    except (FileNotFoundError, json.JSONDecodeError, OSError):
        data = {}
    data.setdefault(domain, []).append({
        "label": label, "summary": summary, "source": source})
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w", encoding="utf-8") as fh:
        json.dump(data, fh, indent=2, sort_keys=True)
    return {"domain": domain, "candidate": label}


def list_candidates(domain: str = "") -> dict:
    """List promotion candidates (all domains, or one domain)."""
    try:
        with open(_candidates_path(), "r", encoding="utf-8") as fh:
            data = json.load(fh)
    except (FileNotFoundError, json.JSONDecodeError, OSError):
        return {}
    if domain:
        return {domain: data.get(domain, [])}
    return data


def _cmd_promote(args) -> int:
    if args.action == "add":
        res = add_candidate(args.domain, args.label, args.summary, args.source)
    elif args.action == "list":
        res = list_candidates(args.domain)
    else:
        print("usage: promote add <domain> <label> <summary> | promote list [domain]")
        return 2
    print(json.dumps(res, indent=2, sort_keys=True, default=str))
    return 0


def estimate_tokens_doctrine(block: str) -> int:
    """Estimate tokens of a doctrine block (4 chars ≈ 1 token)."""
    return retrieval.estimate_tokens(block)


def _cmd_doctrine(args) -> int:
    block = build_doctrine(args.project_id, args.query, args.budget)
    print(block)
    return 0


def build_parser() -> argparse.ArgumentParser:
    """CLI: vitriol pymander <cmd>. Subcommands mirror the P1 scope."""
    ap = argparse.ArgumentParser(prog="vitriol pymander",
                                 description="Pymander reference-mind store")
    sub = ap.add_subparsers(dest="cmd", required=True)

    p_list = sub.add_parser("list", help="list installed Pymander domains")
    p_list.set_defaults(fn=_cmd_list)

    p_ingest = sub.add_parser(
        "ingest", help="ingest a markdown corpus as atomic nodes")
    p_ingest.add_argument("domain")
    p_ingest.add_argument("file", help="corpus .md, or '-' for stdin")
    p_ingest.add_argument("--rev", default=None,
                          help="git_rev; default: corpus repo HEAD ('' outside git)")
    p_ingest.set_defaults(fn=_cmd_ingest)

    p_nodes = sub.add_parser("nodes", help="list a domain's current nodes")
    p_nodes.add_argument("domain")
    p_nodes.set_defaults(fn=_cmd_nodes)

    p_search = sub.add_parser("search", help="retrieve nodes for a query")
    p_search.add_argument("domain")
    p_search.add_argument("query")
    p_search.add_argument("--top-k", type=int, default=5)
    p_search.set_defaults(fn=_cmd_search)

    p_sel = sub.add_parser(
        "select", help="set the active domains for a project")
    p_sel.add_argument("project_id")
    p_sel.add_argument("domains", nargs="+")
    p_sel.set_defaults(fn=_cmd_select)

    p_active = sub.add_parser(
        "active", help="show the active domains for a project")
    p_active.add_argument("project_id")
    p_active.set_defaults(fn=_cmd_active)

    p_doctrine = sub.add_parser(
        "doctrine", help="build a budgeted doctrine block for a project")
    p_doctrine.add_argument("project_id")
    p_doctrine.add_argument("--query", default="")
    p_doctrine.add_argument("--budget", type=int, default=3000)
    p_doctrine.set_defaults(fn=_cmd_doctrine)

    p_promote = sub.add_parser(
        "promote", help="curate Ascensus/Hermetis answers into candidates")
    p_promote.add_argument("action", choices=["add", "list"])
    p_promote.add_argument("domain", nargs="?")
    p_promote.add_argument("label", nargs="?")
    p_promote.add_argument("summary", nargs="?")
    p_promote.add_argument("--source", default="")
    p_promote.set_defaults(fn=_cmd_promote)
    return ap


def main(argv=None) -> int:
    """CLI entry point (also usable from scripts/vitriol)."""
    args = build_parser().parse_args(argv)
    return args.fn(args)


if __name__ == "__main__":
    sys.exit(main())

File: libhermes/src/pymander.rs
//! Pymander — the reference mind: static, curated, domain-specific knowledge.
//!
//! Port of `libvitriol/pymander.py`. Each domain is a distinct Hermetis memory
//! root `~/.vitriol/pymander/<domain>/memory.db` (project_id =
//! `pymander/<domain>`), so node versioning, the embedding cache, and strength
//! come from the existing db machinery. Ingest format: `## Heading` starts an
//! atomic node; the body up to the next `##` is its summary.

use std::path::{Path, PathBuf};

use regex::Regex;

use crate::db::Hermes;
use crate::retrieval::{retrieve, RetrieveParams};
use crate::scorer::estimate_tokens;
use crate::NodeMeta;

const DOMAIN_PREFIX: &str = "pymander/";

/// Lazy domain-name validator: `^[A-Za-z0-9][A-Za-z0-9._-]*$`.
fn domain_re() -> &'static Regex {
    static RE: std::sync::OnceLock<Regex> = std::sync::OnceLock::new();
    RE.get_or_init(|| Regex::new(r"^[A-Za-z0-9][A-Za-z0-9._-]*$").unwrap())
}

/// Validate + normalize a domain name (raises a string error).
pub fn sanitize_domain(domain: &str) -> Result<String, String> {
    if domain_re().is_match(domain) {
        Ok(domain.to_string())
    } else {
        Err(format!(
            "invalid domain {domain:?}: use letters/digits/._- , no slashes"
        ))
    }
}

/// The Hermetis project_id backing a Pymander domain.
pub fn domain_project_id(domain: &str) -> Result<String, String> {
    Ok(format!("{DOMAIN_PREFIX}{}", sanitize_domain(domain)?))
}

fn selection_path(memory_root: &Path) -> PathBuf {
    memory_root.join("pymander/selection.json")
}

fn candidates_path(memory_root: &Path) -> PathBuf {
    memory_root.join("pymander/candidates.json")
}

/// Installed Pymander domains (dirs holding a memory.db), sorted.
pub fn list_domains(memory_root: &Path) -> Vec<String> {
    let root = memory_root.join("pymander");
    let Ok(entries) = std::fs::read_dir(&root) else {
        return Vec::new();
    };
    let mut out = Vec::new();
    for e in entries.flatten() {
        let p = e.path();
        if p.is_dir() && p.join("memory.db").exists() {
            if let Some(name) = p.file_name() {
                out.push(name.to_string_lossy().into_owned());
            }
        }
    }
    out.sort();
    out
}

/// Split markdown into (label, summary) atomic nodes — Python `_parse_markdown`.
pub fn parse_markdown(md_text: &str) -> Vec<(String, String)> {
    let mut nodes = Vec::new();
    let mut label: Option<String> = None;
    let mut body: Vec<String> = Vec::new();
    for line in md_text.lines() {
        if let Some(rest) = line.strip_prefix("## ") {
            flush_node(&mut nodes, label.take(), &mut body);
            label = Some(rest.trim().to_string());
        } else if label.is_some() {
            body.push(line.to_string());
        }
    }
    flush_node(&mut nodes, label, &mut body);
    nodes
}

fn flush_node(nodes: &mut Vec<(String, String)>, label: Option<String>, body: &mut Vec<String>) {
    if let Some(l) = label {
        let summary = body.join("\n").trim().to_string();
        if !summary.is_empty() {
            nodes.push((l, summary));
        }
        body.clear();
    }
}

/// Ingest a markdown corpus as atomic nodes for a domain.
pub fn ingest_markdown(
    h: &Hermes,
    domain: &str,
    md_text: &str,
    git_rev: &str,
) -> Result<serde_json::Value, String> {
    let domain = sanitize_domain(domain)?;
    let pid = domain_project_id(&domain)?;
    let nodes = parse_markdown(md_text);
    let conn = h.conn(&pid).map_err(|e| e.to_string())?;
    let mut stored = 0usize;
    let mut refreshed = 0usize;
    for (label, summary) in &nodes {
        let exists = conn
            .query_row(
                "SELECT 1 FROM knowledge_nodes WHERE label=?1 AND git_rev=?2",
                rusqlite::params![label, git_rev],
                |_| Ok(()),
            )
            .is_ok();
        h.store_node(
            &pid,
            label,
            summary,
            &NodeMeta {
                git_rev: git_rev.to_string(),
                strength: 1.0,
                ..Default::default()
            },
        )
        .map_err(|e| e.to_string())?;
        if exists {
            refreshed += 1;
        } else {
            stored += 1;
        }
    }
    Ok(serde_json::json!({
        "domain": domain,
        "nodes": nodes.len(),
        "stored": stored,
        "refreshed": refreshed,
        "embedded": 0, // semantic-off: no embed provider
    }))
}

/// Current (superseded=0) nodes of a domain.
pub fn list_nodes(h: &Hermes, domain: &str) -> Result<Vec<serde_json::Value>, String> {
    let pid = domain_project_id(domain)?;
    h.fetch_nodes(&pid, 100_000, false)
        .map_err(|e| e.to_string())
}

/// The most relevant nodes of a domain for a query.
pub fn search(
    h: &Hermes,
    domain: &str,
    query: &str,
    top_k: usize,
) -> Result<Vec<serde_json::Value>, String> {
    let pid = domain_project_id(domain)?;
    let params = RetrieveParams {
        top_k,
        cascade_depth: 0,
        include_history: false,
        candidate_multiplier: 10,
    };
    let candidates = retrieve(h, &pid, query, &params);
    Ok(candidates
        .into_iter()
        .filter(|c| c.get("_type").and_then(|t| t.as_str()) == Some("node"))
        .collect())
}

/// Persisted selection.json read (missing/corrupt -> empty).
fn load_selection(memory_root: &Path) -> serde_json::Value {
    let path = selection_path(memory_root);
    std::fs::read_to_string(&path)
        .ok()
        .and_then(|t| serde_json::from_str(&t).ok())
        .filter(|v: &serde_json::Value| v.is_object())
        .unwrap_or_else(|| serde_json::json!({}))
}

/// Atomically write selection.json.
fn save_selection(memory_root: &Path, data: &serde_json::Value) -> Result<(), String> {
    let path = selection_path(memory_root);
    if let Some(dir) = path.parent() {
        std::fs::create_dir_all(dir).map_err(|e| e.to_string())?;
    }
    let text = serde_json::to_string_pretty(data).map_err(|e| e.to_string())?;
    let tmp = path.with_extension("json.tmp");
    std::fs::write(&tmp, text).map_err(|e| e.to_string())?;
    std::fs::rename(&tmp, &path).map_err(|e| e.to_string())
}

/// Set the active Pymander domains for a project (persisted).
pub fn set_selection(
    memory_root: &Path,
    project_id: &str,
    domains: &[String],
) -> Result<serde_json::Value, String> {
    let mut data = load_selection(memory_root);
    let clean: Vec<String> = domains
        .iter()
        .map(|d| sanitize_domain(d))
        .collect::<Result<_, _>>()?;
    data[project_id] = serde_json::json!(clean);
    save_selection(memory_root, &data)?;
    Ok(serde_json::json!({"project_id": project_id, "domains": clean}))
}

/// The active Pymander domains for a project (empty if unset).
pub fn get_selection(memory_root: &Path, project_id: &str) -> Vec<String> {
    load_selection(memory_root)
        .get(project_id)
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|d| d.as_str())
                .map(str::to_string)
                .collect()
        })
        .unwrap_or_default()
}

/// Best-effort git HEAD of the corpus file's repo ('' when not in a repo).
pub fn repo_rev(path: &str) -> String {
    let p = Path::new(path);
    if path == "-" || !p.is_file() {
        return String::new();
    }
    let base = p.parent().unwrap_or(Path::new("."));
    if let Ok(out) = std::process::Command::new("git")
        .arg("-C")
        .arg(base)
        .args(["rev-parse", "HEAD"])
        .output()
    {
        if out.status.success() {
            if let Ok(s) = String::from_utf8(out.stdout) {
                return s.trim().to_string();
            }
        }
    }
    String::new()
}

/// Doctrine builder options.
#[derive(Debug, Clone)]
pub struct DoctrineOpts {
    pub query: String,
    pub budget_tokens: usize,
    pub top_k: usize,
}

/// Build a budgeted doctrine block for the project's selected domains.
pub fn build_doctrine(
    h: &Hermes,
    memory_root: &Path,
    project_id: &str,
    o: &DoctrineOpts,
) -> String {
    let domains = get_selection(memory_root, project_id);
    let domains = if domains.is_empty() {
        list_domains(memory_root).into_iter().take(1).collect()
    } else {
        domains
    };
    let mut sections = Vec::new();
    let mut used = 0usize;
    for domain in domains {
        let text = domain_section(
            h,
            &domain,
            &o.query,
            o.top_k,
            o.budget_tokens.saturating_sub(used),
        );
        if text.is_empty() {
            continue;
        }
        let text_toks = estimate_tokens(&text) + 1;
        if used + text_toks > o.budget_tokens && used > 0 {
            break;
        }
        used += text_toks;
        sections.push(text);
    }
    sections.join("\n\n")
}

/// One domain's doctrine lines under a per-domain budget.
fn domain_section(h: &Hermes, domain: &str, query: &str, top_k: usize, budget: usize) -> String {
    let Ok(hits) = search(h, domain, query, top_k) else {
        return String::new();
    };
    if hits.is_empty() {
        return String::new();
    }
    let mut parts = vec![format!("## {domain}")];
    let mut used = 0usize;
    for hit in &hits {
        let label = hit.get("label").and_then(|l| l.as_str()).unwrap_or("");
        let summary = hit.get("summary").and_then(|s| s.as_str()).unwrap_or("");
        let body = format!("- {label}: {summary}");
        let toks = estimate_tokens(&body) + 1;
        if used + toks > budget && used > 0 {
            break;
        }
        used += toks;
        parts.push(body);
    }
    parts.join("\n")
}

/// Add a promotion candidate (curated, never auto-merged).
pub fn add_candidate(
    memory_root: &Path,
    domain: &str,
    label: &str,
    summary: &str,
    source: &str,
) -> Result<serde_json::Value, String> {
    let domain = sanitize_domain(domain)?;
    let path = candidates_path(memory_root);
    let mut data = std::fs::read_to_string(&path)
        .ok()
        .and_then(|t| serde_json::from_str(&t).ok())
        .filter(|v: &serde_json::Value| v.is_object())
        .unwrap_or_else(|| serde_json::json!({}));
    if data.get(&domain).and_then(|v| v.as_array()).is_none() {
        data[&domain] = serde_json::json!([]);
    }
    if let Some(arr) = data.get_mut(&domain).and_then(|v| v.as_array_mut()) {
        arr.push(serde_json::json!({"label": label, "summary": summary, "source": source}));
    }
    if let Some(dir) = path.parent() {
        std::fs::create_dir_all(dir).map_err(|e| e.to_string())?;
    }
    let text = serde_json::to_string_pretty(&data).map_err(|e| e.to_string())?;
    std::fs::write(&path, text).map_err(|e| e.to_string())?;
    Ok(serde_json::json!({"domain": domain, "candidate": label}))
}

/// List promotion candidates (all domains, or one domain).
pub fn list_candidates(memory_root: &Path, domain: &str) -> serde_json::Value {
    let data = std::fs::read_to_string(candidates_path(memory_root))
        .ok()
        .and_then(|t| serde_json::from_str(&t).ok())
        .unwrap_or_else(|| serde_json::json!({}));
    if domain.is_empty() {
        data
    } else {
        serde_json::json!({domain: data.get(domain).cloned().unwrap_or(serde_json::json!([]))})
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn tmp_home(name: &str) -> (PathBuf, Hermes) {
        let root = std::env::temp_dir().join(format!("hermes_pymander_{name}"));
        let _ = std::fs::remove_dir_all(&root);
        (root.clone(), Hermes::new(&root))
    }

    #[test]
    fn domain_validation() {
        assert_eq!(sanitize_domain("systems").unwrap(), "systems");
        assert_eq!(sanitize_domain("c-firmware_v2").unwrap(), "c-firmware_v2");
        assert!(sanitize_domain("bad/name").is_err());
        assert!(sanitize_domain("-leading").is_err());
    }

    #[test]
    fn parse_markdown_splits_nodes() {
        let md = "# Domain header\nprose skipped\n\n## First\nbody one\n\n## Second\nbody two\n";
        let nodes = parse_markdown(md);
        assert_eq!(nodes.len(), 2);
        assert_eq!(nodes[0].0, "First");
        assert!(nodes[0].1.contains("body one"));
        assert!(nodes[1].1.contains("body two"));
    }

    #[test]
    fn ingest_and_search_roundtrip() {
        let (root, h) = tmp_home("ingest");
        let md = "## widget\nwidgets are fast\n## parser\nparser handles tokens\n";
        let res = ingest_markdown(&h, "systems", md, "abc").unwrap();
        assert_eq!(res["stored"], 2);
        assert_eq!(res["refreshed"], 0);
        // re-ingest same rev -> refreshed
        let res = ingest_markdown(&h, "systems", md, "abc").unwrap();
        assert_eq!(res["stored"], 0);
        assert_eq!(res["refreshed"], 2);
        // search finds the parser node for a token query
        let hits = search(&h, "systems", "token parser", 5).unwrap();
        assert!(!hits.is_empty());
        assert_eq!(hits[0]["label"], "parser");
        let _ = std::fs::remove_dir_all(&root);
    }

    #[test]
    fn selection_and_doctrine() {
        let (root, h) = tmp_home("doctrine");
        let md = "## rust\nownership is key\n";
        ingest_markdown(&h, "rust", md, "abc").unwrap();
        let res = set_selection(&root, "proj", &["rust".to_string()]).unwrap();
        assert_eq!(res["domains"][0], "rust");
        assert_eq!(get_selection(&root, "proj"), vec!["rust".to_string()]);
        let block = build_doctrine(
            &h,
            &root,
            "proj",
            &DoctrineOpts {
                query: "ownership".into(),
                budget_tokens: 1000,
                top_k: 3,
            },
        );
        assert!(block.contains("## rust"));
        assert!(block.contains("ownership"));
        let _ = std::fs::remove_dir_all(&root);
    }
}
